File: 1156. Swap For Longest Repeated Character Substring.cpp
class Solution {
public:
    int maxRepOpt1(string s) {
        int n = s.size();
        vector<int> cnt(129), f(n);
        for (auto c : s) ++cnt[c];
        f[0] = 1;
        int ans = 0;
        for (int i=1; i<n; ++i) {
            f[i] = (s[i]==s[i-1] ? f[i-1] : 0) + 1;
            ans = max(ans, f[i]);
            if (i-f[i]==0 && cnt[s[i]]>f[i] || 
                i-f[i]>0 && s[i-f[i]-1]==s[i] && f[i]+f[i-f[i]-1]<cnt[s[i]])
                ans = max(ans, f[i]+1+(i-f[i]==0 ? 0 : f[i-f[i]-1]));
            if (s[i]!=s[i-1] && cnt[s[i-1]]>f[i-1])
                ans = max(ans, f[i-1]+1);
        }
        return ans;
    }
};

File: 1156. Swap For Longest Repeated Character Substring.py
class Solution:
    def maxRepOpt1(self, s: str) -> int:
        n = len(s)
        c = collections.Counter(s)
        a = [[x, len(list(it))] for x,it in itertools.groupby(s)]
        ans = max(min(c[x], l+1) for x,l in a)
        for i in range(1, len(a)-1):
            if a[i-1][0]==a[i+1][0] and a[i][1]==1:
                ans = max(ans, min(a[i-1][1]+a[i+1][1]+1, c[a[i-1][0]]))
        return ans
        